# Dataframe Exercises

### For several of the following exercises, you'll need to load several datasets using the pydataset library. (If you get an error when trying to run the import below, use pip to install the pydataset package.)

```from pydataset import data```



In [1]:
from pydataset import data

initiated datasets repo at: /Users/charlesatchison/.pydataset/


### Copy the code from the lesson to create a dataframe full of studen grades.

In [41]:
import pandas as pd
import numpy as np

np.random.seed(123)

students = ['Sally', 'Jane', 'Suzie', 'Billy', 'Ada', 'John', 'Thomas',
            'Marie', 'Albert', 'Richard', 'Isaac', 'Alan']

# randomly generate scores for each student for each subject
# note that all the values need to have the same length here
math_grades = np.random.randint(low=60, high=100, size=len(students))
english_grades = np.random.randint(low=60, high=100, size=len(students))
reading_grades = np.random.randint(low=60, high=100, size=len(students))

df = pd.DataFrame({'name': students,
                   'math': math_grades,
                   'english': english_grades,
                   'reading': reading_grades})
df

name  math  english  reading
0     Sally    62       85       80
1      Jane    88       79       67
2     Suzie    94       74       95
3     Billy    98       96       88
4       Ada    77       92       98
5      John    79       76       93
6    Thomas    82       64       81
7     Marie    93       63       90
8    Albert    92       62       87
9   Richard    69       80       94
10    Isaac    92       99       93
11     Alan    92       62       72

### 1a. Create a column named ```passing_english``` that indicates whether each student has a passing grade in english.

In [42]:
df['passing_english'] = df.english > 70
df

name  math  english  reading  passing_english
0     Sally    62       85       80             True
1      Jane    88       79       67             True
2     Suzie    94       74       95             True
3     Billy    98       96       88             True
4       Ada    77       92       98             True
5      John    79       76       93             True
6    Thomas    82       64       81            False
7     Marie    93       63       90            False
8    Albert    92       62       87            False
9   Richard    69       80       94             True
10    Isaac    92       99       93             True
11     Alan    92       62       72            False

### 1b. Sort the english grades by the ```passing_english``` column. How are duplicates handled?


#### The duplicates are handled by sorting by the index value next

In [43]:
df.sort_values(by='passing_english')

name  math  english  reading  passing_english
6    Thomas    82       64       81            False
7     Marie    93       63       90            False
8    Albert    92       62       87            False
11     Alan    92       62       72            False
0     Sally    62       85       80             True
1      Jane    88       79       67             True
2     Suzie    94       74       95             True
3     Billy    98       96       88             True
4       Ada    77       92       98             True
5      John    79       76       93             True
9   Richard    69       80       94             True
10    Isaac    92       99       93             True

### 1c. Sort the english grades first by passing_english and then by student name. All the students that are failing english should be first, and within the students that are failing english they should be ordered alphabetically. The same should be true for the students passing english. (Hint: you can pass a list to the .sort_values method)

In [44]:
df.sort_values(by=['passing_english', 'name'])

name  math  english  reading  passing_english
11     Alan    92       62       72            False
8    Albert    92       62       87            False
7     Marie    93       63       90            False
6    Thomas    82       64       81            False
4       Ada    77       92       98             True
3     Billy    98       96       88             True
10    Isaac    92       99       93             True
1      Jane    88       79       67             True
5      John    79       76       93             True
9   Richard    69       80       94             True
0     Sally    62       85       80             True
2     Suzie    94       74       95             True

### 1d. Sort the english grades first by ```passing_english```, and then by the actual english grade, similar to how we did in the last step.

In [45]:
df.sort_values(by=['passing_english', 'english'])

name  math  english  reading  passing_english
8    Albert    92       62       87            False
11     Alan    92       62       72            False
7     Marie    93       63       90            False
6    Thomas    82       64       81            False
2     Suzie    94       74       95             True
5      John    79       76       93             True
1      Jane    88       79       67             True
9   Richard    69       80       94             True
0     Sally    62       85       80             True
4       Ada    77       92       98             True
3     Billy    98       96       88             True
10    Isaac    92       99       93             True

### 1e. Calculate each students overall grade and add it as a column on the dataframe. The overall grade is the average of the math, english, and reading grades.

In [46]:
df['overall_grade'] = df[['math', 'english', 'reading']].mean(axis=1)

name  math  english  reading  passing_english  overall_grade
0     Sally    62       85       80             True      75.666667
1      Jane    88       79       67             True      78.000000
2     Suzie    94       74       95             True      87.666667
3     Billy    98       96       88             True      94.000000
4       Ada    77       92       98             True      89.000000
5      John    79       76       93             True      82.666667
6    Thomas    82       64       81            False      75.666667
7     Marie    93       63       90            False      82.000000
8    Albert    92       62       87            False      80.333333
9   Richard    69       80       94             True      81.000000
10    Isaac    92       99       93             True      94.666667
11     Alan    92       62       72            False      75.333333

## 2. Load the ```mpg``` dataset. Read the documentation for the dataset and use it for the following questions:

In [50]:
mpg = data('mpg')


### 2a. How many rows and columns are there? -- 

234 rows, 11 columns

In [51]:
mpg.shape

(234, 11)

### 2b. What are the data types of each column?

In [54]:
mpg.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 234 entries, 1 to 234
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   manufacturer  234 non-null    object 
 1   model         234 non-null    object 
 2   displ         234 non-null    float64
 3   year          234 non-null    int64  
 4   cyl           234 non-null    int64  
 5   trans         234 non-null    object 
 6   drv           234 non-null    object 
 7   cty           234 non-null    int64  
 8   hwy           234 non-null    int64  
 9   fl            234 non-null    object 
 10  class         234 non-null    object 
dtypes: float64(1), int64(4), object(6)
memory usage: 21.9+ KB


### 2c. Summarize the dataframe with ```.info``` and ```.describe```

In [55]:
mpg.info(), mpg.describe()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 234 entries, 1 to 234
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   manufacturer  234 non-null    object 
 1   model         234 non-null    object 
 2   displ         234 non-null    float64
 3   year          234 non-null    int64  
 4   cyl           234 non-null    int64  
 5   trans         234 non-null    object 
 6   drv           234 non-null    object 
 7   cty           234 non-null    int64  
 8   hwy           234 non-null    int64  
 9   fl            234 non-null    object 
 10  class         234 non-null    object 
dtypes: float64(1), int64(4), object(6)
memory usage: 21.9+ KB


(None,
             displ         year         cyl         cty         hwy
 count  234.000000   234.000000  234.000000  234.000000  234.000000
 mean     3.471795  2003.500000    5.888889   16.858974   23.440171
 std      1.291959     4.509646    1.611534    4.255946    5.954643
 min      1.600000  1999.000000    4.000000    9.000000   12.000000
 25%      2.400000  1999.000000    4.000000   14.000000   18.000000
 50%      3.300000  2003.500000    6.000000   17.000000   24.000000
 75%      4.600000  2008.000000    8.000000   19.000000   27.000000
 max      7.000000  2008.000000    8.000000   35.000000   44.000000)

### 2d. Rename the `cty` coliumn to `city`.

In [56]:
mpg.rename(columns={'cty': 'city'})

manufacturer   model  displ  year  cyl       trans drv  city  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f    18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f    21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f    20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f    21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f    16   26  p   
..           ...     ...    ...   ...  ...         ...  ..   ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f    19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f    21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f    16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f    18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f    17   26  p   

       class  
1    compact  
2    compact  
3    compact  
4    compact  
5    compact  
..       ...  
230  midsize  
231  midsize  
232  midsize  
233  midsize  
234  midsize  

[234 rows x 11 columns]

### 2e. Rename `hwy` column to `highway`.

In [57]:
mpg.rename(columns={'hwy': 'highway'})

manufacturer   model  displ  year  cyl       trans drv  cty  highway fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18       29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21       29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20       31  p   
4           audi      a4    2.0  2008    4    auto(av)   f   21       30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16       26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...      ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19       28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21       29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16       26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18       26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17       26  p   

       class  
1    compact  
2    compact  
3    compact  
4    compact  
5    compact  
..       ...  
230  midsize  
231  midsize  
232  midsize  
233  midsize  
234  midsize  

[234 rows x 11 columns]

### 2f. Do any cars have better city mileage than highway mileage? -- 
No

In [68]:
mpg[mpg['cty'] > mpg['hwy']]

Empty DataFrame
Columns: [manufacturer, model, displ, year, cyl, trans, drv, cty, hwy, fl, class]
Index: []

### 2g. Create a column named `mileage_difference` this column should contain the difference between highway and city mileage for each car.

In [76]:
mpg['mileage_difference'] = mpg['hwy'] - mpg['cty']
mpg

manufacturer   model  displ  year  cyl       trans drv  cty  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f   21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16   26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26  p   

       class  mileage_difference  
1    compact                  11  
2    compact                   8  
3    compact                  11  
4    compact                   9  
5    compact                  10  
..       ...                 ...  
230  midsize                   9  
231  midsize                   8  
232  midsize                  10  
233  midsize                   8  
234  midsize                   9  

[234 rows x 12 columns]

### 2h. Which car (or cars) has the highest mileage difference?

In [86]:
mpg[mpg.mileage_difference == mpg.mileage_difference.max()]

manufacturer       model  displ  year  cyl     trans drv  cty  hwy fl  \
107        honda       civic    1.8  2008    4  auto(l5)   f   24   36  c   
223   volkswagen  new beetle    1.9  1999    4  auto(l4)   f   29   41  d   

          class  mileage_difference  
107  subcompact                  12  
223  subcompact                  12

### 2i. Which compact class car has the lowest highway mileage? The best?

In [131]:
cc = mpg[mpg['class'] == 'compact']
max_min = cc.hwy.max(), cc.hwy.min()
hgst_hwy = cc[cc.hwy == max_min[0]]
lwst_hwy = cc[cc.hwy == max_min[1]]

manufacturer model  displ  year  cyl trans  drv  cty  hwy   fl class  \
213          NaN   NaN    NaN   NaN  NaN   NaN  NaN  NaN  NaN  NaN   NaN   
220          NaN   NaN    NaN   NaN  NaN   NaN  NaN  NaN  NaN  NaN   NaN   

     mileage_difference  average_mileage  
213                 NaN              NaN  
220                 NaN              NaN

### 2j. Create a column named `average_mileage` that is the mean of the city and highway mileage.

In [103]:
mpg['average_mileage'] = mpg[['cty', 'hwy']].mean(axis=1)

### 2k. Which Dodge car has the best average mileage? The worst?

In [132]:
dodge_vehicles = mpg[mpg.manufacturer == 'dodge']
max_min = dodge_vehicles.average_mileage.max(), dodge_vehicles.average_mileage.min()
bst = dodge_vehicles[dodge_vehicles.average_mileage == max_min[0]]
wrst = dodge_vehicles[dodge_vehicles.average_mileage == max_min[1]]

## 3. Load the `Mammals` dataset. Read the documentation for it, and use the data to answer these questions:

In [133]:
mammals = data('Mammals')

### 3a. How many rows and columns are there? 107 Rows, 4 Columns

In [134]:
mammals.shape

(107, 4)

### 3b. What are the data types?

In [136]:
mammals.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 107 entries, 1 to 107
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   weight    107 non-null    float64
 1   speed     107 non-null    float64
 2   hoppers   107 non-null    bool   
 3   specials  107 non-null    bool   
dtypes: bool(2), float64(2)
memory usage: 2.7 KB


### 3c. Summarize the dataframe with `.info` and `.describe`.

In [137]:
mammals.info(), mammals.describe()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 107 entries, 1 to 107
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   weight    107 non-null    float64
 1   speed     107 non-null    float64
 2   hoppers   107 non-null    bool   
 3   specials  107 non-null    bool   
dtypes: bool(2), float64(2)
memory usage: 2.7 KB


(None,
             weight       speed
 count   107.000000  107.000000
 mean    278.688178   46.208411
 std     839.608269   26.716778
 min       0.016000    1.600000
 25%       1.700000   22.500000
 50%      34.000000   48.000000
 75%     142.500000   65.000000
 max    6000.000000  110.000000)

### 3d. What is the weight of the fastest animal? 55

In [146]:
top_spd = mammals.speed.max()
mammals[mammals.speed == top_spd].weight

53    55.0
Name: weight, dtype: float64

### 3e. What is the overall percentage of specials?

In [158]:
print(f'{(mammals.specials.sum()/len(mammals.specials)*100):0.2f}' + ' %')

9.35 %


### 3f. How many animals are hoppers that are above the median speed? What percentage is this?

In [176]:
hoppers = mammals[mammals.hoppers]
fst_hoppers = (hoppers.speed > mammals.speed.mean()).sum()
prct_hop = fst_hoppers / len(hoppers) * 100
prct_ttl = fst_hoppers / len(mammals * 100)
print(f'Percentage of hoppers that are above the median speed: {prct_hop:0.2f}%')
print(f'Percentage of hoppers (relative to all mammals) that are above the median speed: {prct_ttl:0.2f}%')


Percentage of hoppers that are above the median speed: 63.64%
Percentage of hoppers (relative to all mammals) that are above the median speed: 0.07%
